### ISSUE TITLE
Using null conditional in assertion should be reported as code smell

### ISSUE THREAD

**Issue by krajek** [fluentassertions/fluentassertions.analyzers#21]
In main FA repository, there was an issue reported https://github.com/fluentassertions/fluentassertions/issues/569.

To keep it short, the user used `?.` operator:

```
int? val = null;
val?.Should().Be(400);
```

instead of regular

```
val.Should().Be(400);
```

Do you think it is possible to detect such scenario and suggest code fix?

**Comment by jnyrup** [comment]
That is not only a code smell for `Nullable` but for all reference types where null conditional operator is used before calling `Should()`

In this example `Should()` is never called, as the null conditional operator short-circuits the expression.
```
string str = null;
str?.Should().NotBeNull();
```

There is [`SyntaxKind.ConditionalAccessExpression`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs#L359) to check for usages of `a?.b` and `a?[1]`.

**Comment by Meir017** [comment]
I have a working sample for this.
![image](https://user-images.githubusercontent.com/9786571/34838520-110df70c-f707-11e7-94ea-915ecd55f640.png)

```cs
class ConditionalAccessExpressionVisitor : CSharpSyntaxWalker
{
    public bool CodeSmells { get; private set; }
    public Location ConditionalAccess { get; private set; }

    public override void VisitConditionalAccessExpression(ConditionalAccessExpressionSyntax node)
    {
        if (CodeSmells) return;

        CodeSmells = node.WhenNotNull is InvocationExpressionSyntax invocation
            && invocation.Expression is MemberAccessExpressionSyntax memberAccess && memberAccess.IsKind(SyntaxKind.SimpleMemberAccessExpression)
            && memberAccess.Expression is InvocationExpressionSyntax shouldInvocation
            && shouldInvocation.Expression is MemberBindingExpressionSyntax memberBinding
            && memberBinding.Name.Identifier.ValueText == "Should";
        if (CodeSmells)
        {
            ConditionalAccess = node.GetLocation();
        }
    }
}
```
